### PR TITLE
Nextup and nextdown functions

### PR DESCRIPTION
**Pull request by pdrocaldeira**
nextupN and nextdownN implementation of TS 18661-2.

It's a nextafter but always towards positive/negative infinity

Related to #94

### CODE REVIEW

File: tests/nextdown.input [hunk 1]
```diff
@@ -0,0 +1,31 @@
+# name nextdown
+# arg1 decimal
+# ret  decimal
+
+0                                                   -0.000001 decimal32
```
Comment (pmur) [inline tests/nextdown.input]: Should this be returning -subnormal_min for all formats? e.g 0.000001e-DEC_MIN_EXP.

nextAfter may not be handling subnormals correctly.

Comment (pdrocaldeira) [reply tests/nextdown.input]: When the argument is 0 I think it makes sense the result being the smallest number possible.

But for normal numbers you can't increment subnormal values because you will need to change the number representation to have the smallest exponent possible and that will overflow the mantissa.

If you see in ISO-IEC TR 24732 2008-05-16 they have this `EPSILON` concept and I think they are using it for every number, zero included.
>the difference between 1 and the least value greater than 1 that is representable in the given floating point type:
DEC32_EPSILON  1E-6DF 
DEC64_EPSILON  1E-15DD 
DEC128_EPSILON 1E-33DL

If that is not the case and `nextAfter` is really broken I can create another PR and fix `nextAfter` before merging this PR.

Comment (pmur) [reply tests/nextdown.input]: I believe format in this context implies decimal{32,64,128}. This function may or may not change the quantum exponent.  This is an implementation bug. subnormals make for tricky edge cases.

**Comment by pdrocaldeira** [comment]
These last commits broke `test-exp-d128` and I don't have a clue about it. :cry: 

```
Failure: Test: exp (-1414400000000000000000000000000000e-29)
Result:
 is:          2.18196590654201125512e-6143  2.181965906542011255118719873241674e-6143
 should be:   0.00000000000000000000e+00  0.000000000000000000000000000000000e+00
 difference:  2.18196590654201125512e-6143  2.181965906542011255118719873241674e-6143
 ulp       :  2181965906542011255118719873241674.0000
 max.ulp   :  1.0000
```

Now `nextafter` is a native function and is returning subnormal values for zero and becoming zero if it's going forward in the opposite direction of its signal.

**Comment by pmur** [comment]
Good job catching a bug in the rounding of the (my) auto-dfp test generator! It incorrectly rounded the result to 0. The ULPs calculation looks correct.

**Comment by pmur** [comment]
I think you can safely remove the FUNC(fpclassify)() tests in tests/libdfp-test.c how that nextafter is behaving correctly.

File: ieee754r/nextafterd32.c [hunk 2]
```diff
@@ -26,63 +26,41 @@
...
+  {
+    if(x == DEC_INFINITY) return DEC_MAX;
+    epsilon *= -1;
+  }
+  else if(x == (-1*DEC_INFINITY)) return -1*DEC_MAX;
+  if (x == 0) return epsilon;
+  DEC_TYPE justified = FUNC_D(left_justify) (x);
+  int exponent = FUNC_D (getexp) (justified);
+  epsilon = FUNC_D (setexp) (epsilon, exponent);
+  return x + epsilon;
```
Comment (pmur) [inline ieee754r/nextafterd32.c]: One minor request, can you pass this through indent to more closely match GNU coding standards?

**Comment by pdrocaldeira** [comment]
I think now everything is ok.
I've passed `indent` on next after file and now we have nextup/nextdown working properly even with subnormal numbers. You can see that more objectively looking at the updated values on left-justify tests cases. Code was tested on `ppc64le` and `x86_64`.

The problem is that the left-justify Power function was not respecting the minimum exponent. So it was propagating a NaN value. Now if we need to go to a lower exponent we respect the biased limit (0).

File: sysdeps/powerpc/dfpu/numdigits.h [hunk 3]
```diff
@@ -185,7 +185,8 @@ FUNC_D (left_justify) (DEC_TYPE x)
   asm ("dxex" Q " %0,%1\n\t" : "=d"(tmp2) : "d"(rnd));
 
   d2.f = tmp2;
-  d2.l -= ADJUST;
+  if(d2.l >= ADJUST) d2.l -= ADJUST;
```
Comment (pmur) [inline sysdeps/powerpc/dfpu/numdigits.h]: Minor nit, formatting. Can you run this file through indent? Changes should be formatted in GNU style.

Likewise, could you add some extra tests to numdigits to verify subnormals behave? I don't think _Decimal32 will ever have issues on ppc, but it may be worth the effort for other platforms (e.g s390 or x86) which also implement this in a platform specific manner.

File: sysdeps/soft-dfp/bid/numdigits.h [hunk 4]
```diff
@@ -371,31 +371,37 @@ static inline _Decimal64
...
+
 	  /* If the number overflows the data it will become a NaN1. */
-	  if ((exp - firstdigit) <= -DECIMAL64_Bias && firstdigit != 0)
+	  if (leadingzeroes < 0)
 	    {
-	      union ieee754r_Decimal64 mask = { .di = { 0x00000001,
-                                                        0x7c000000 } };
+	      union ieee754r_Decimal64 mask = {.di = {0x00000001,
+						      0x7c000000}
+	      };
```
Comment (pmur) [inline sysdeps/soft-dfp/bid/numdigits.h]: I would recommend adding a test case (if one is not present yet) to validate this condition for all decimal types.

Comment (pdrocaldeira) [reply sysdeps/soft-dfp/bid/numdigits.h]: Nice, thanks to call my attention on that. 
I did the math and this case was never going to happen.
What I did is modify the function to check if the number is NaN right in the beginning and return the number. 
https://github.com/libdfp/libdfp/blob/1388913c5b5cd8b1805cea99379ef98334c7135b/sysdeps/soft-dfp/bid/numdigits.h#L344

File: sysdeps/soft-dfp/bid/numdigits.h [hunk 5]
```diff
@@ -371,31 +371,37 @@ static inline _Decimal64
...
-  char digits[NDIGITS_64+NDIGITS_64-1] = { 0 };
+  char digits[NDIGITS_64 + NDIGITS_64 - 1] = { 0 };
 
   __get_digits_d64 (x, digits, NULL, NULL, NULL, NULL);
   while (digits[firstdigit] == '0')
     firstdigit++;
   len = strlen (digits + firstdigit);
   if (len)
     {
+      int leadingzeroes = -1;
```
[the remark this answers is not in the document]
Comment (pdrocaldeira) [reply sysdeps/soft-dfp/bid/numdigits.h]: Sure, the three functions now should have the same behaviour and there's a new test for NaN in all of them :)

File: ieee754r/Versions [hunk 6]
```diff
@@ -264,6 +264,12 @@ libdfp {
     quantumd128;
   }
   LIBDFP_1.0.15 {
+    nextdownd32;
+    nextdownd64;
+    nextdownd128;
+    nextupd32;
```
Comment (pmur) [inline ieee754r/Versions]: check-abi-libdfp fails due to the ordering of these. Please update and double check.

**Comment by pmur** [comment]
Testing on ppc64le I received the following:
```
cat ./check-abi-libdfp.out 
--- ../sysdeps/powerpc/powerpc64/dfpu/libdfp.abilist	2020-01-22 17:17:57.207757770 -0600
+++ libdfp.symlist	2020-01-22 17:18:32.870638938 -0600
@@ -326,3 +325,0 @@ LIBDFP_1.0.15
- nextupd128 F
- nextupd32 F
- nextupd64 F
@@ -331,0 +329,3 @@ LIBDFP_1.0.15
+ nextupd128 F
+ nextupd32 F
+ nextupd64 F
cat ./test-exp-d128.out 
cat ./test-left_justify.out 
25  Error:   Expected: "+0000001e-101"
             Result:   "+0000000e-101"
    in: ../tests/test-left_justify.c:113.

42  Error:   Expected: "+0000000000000001e-398"
             Result:   "+0000000000000000e-398"
    in: ../tests/test-left_justify.c:156.

58  Error:   Expected: "+0000000000000000000000000000000001e-6176"
             Result:   "+0000000000000000000000000000000000e-6176"
    in: ../tests/test-left_justify.c:192.

Found 3 failures.
cat ./test-quantize-d64.out 
FAIL: check-abi-libdfp
FAIL: test-exp-d128
FAIL: test-left_justify
FAIL: test-quantize-d64
Summary of test results:
      4 FAIL
    163 PASS
```

**Comment by pmur** [comment]
ppc64
```
cat ./check-abi-libdfp.out 
--- ../sysdeps/powerpc/powerpc64/dfpu/libdfp.abilist	2020-01-22 17:21:43.112898975 -0600
+++ libdfp.symlist	2020-01-22 17:22:15.663500729 -0600
@@ -326,3 +325,0 @@ LIBDFP_1.0.15
- nextupd128 F
- nextupd32 F
- nextupd64 F
@@ -331,0 +329,3 @@ LIBDFP_1.0.15
+ nextupd128 F
+ nextupd32 F
+ nextupd64 F
cat ./test-exp-d128.out 
cat ./test-left_justify.out 
25  Error:   Expected: "+0000001e-101"
             Result:   "+0000000e-101"
    in: ../tests/test-left_justify.c:113.

42  Error:   Expected: "+0000000000000001e-398"
             Result:   "+0000000000000000e-398"
    in: ../tests/test-left_justify.c:156.

58  Error:   Expected: "+0000000000000000000000000000000001e-6176"
             Result:   "+0000000000000000000000000000000000e-6176"
    in: ../tests/test-left_justify.c:192.

Found 3 failures.
cat ./test-quantize-d64.out 
FAIL: check-abi-libdfp
FAIL: test-exp-d128
FAIL: test-left_justify
FAIL: test-quantize-d64
Summary of test results:
      4 FAIL
    163 PASS
make: *** [tests] Error 1
```

**Comment by pmur** [comment]
ppc32 soft-dfp:
```
cat ./check-abi-libdfp.out 
--- ../sysdeps/powerpc/powerpc32/nodfpu/libdfp.abilist	2020-01-22 17:21:43.112898975 -0600
+++ libdfp.symlist	2020-01-22 17:24:29.155966477 -0600
@@ -314,3 +313,0 @@ LIBDFP_1.0.15
- nextupd128 F
- nextupd32 F
- nextupd64 F
@@ -319,0 +317,3 @@ LIBDFP_1.0.15
+ nextupd128 F
+ nextupd32 F
+ nextupd64 F
cat ./test-left_justify.out 
25  Error:   Expected: "+0000001e-101"
             Result:   "+0000000e-101"
    in: ../tests/test-left_justify.c:113.

32  Error:   Expected: "+0000010000000000e-398"
             Result:   "+0000000000000001e-398"
    in: ../tests/test-left_justify.c:136.

33  Error:   Expected: "+0000100000000000e-398"
             Result:   "+0000000000000001e-398"
    in: ../tests/test-left_justify.c:137.

34  Error:   Expected: "+0001000000000000e-398"
             Result:   "+0000000000000001e-398"
    in: ../tests/test-left_justify.c:138.

35  Error:   Expected: "+0010000000000000e-398"
             Result:   "+0000000000000001e-398"
    in: ../tests/test-left_justify.c:139.

42  Error:   Expected: "+0000000000000001e-398"
             Result:   "+0000000000000000e-398"
    in: ../tests/test-left_justify.c:156.

52  Error:   Expected: "+0000000000000100000000000000000000e-6176"
             Result:   "+0000000000000000000000000000000001e-6176"
    in: ../tests/test-left_justify.c:182.

53  Error:   Expected: "+0000000000000010000000000000000000e-6176"
             Result:   "+0000000000000000000000000000000001e-6176"
    in: ../tests/test-left_justify.c:183.

54  Error:   Expected: "+0000000000000100000000000000000000e-6176"
             Result:   "+0000000000000000000000000000000001e-6176"
    in: ../tests/test-left_justify.c:184.

55  Error:   Expected: "+0000000000000000100000000000000000e-6176"
             Result:   "+0000000000000000000000000000000001e-6176"
    in: ../tests/test-left_justify.c:185.

58  Error:   Expected: "+0000000000000000000000000000000001e-6176"
             Result:   "+0000000000000000000000000000000000e-6176"
    in: ../tests/test-left_justify.c:192.

Found 11 failures.
FAIL: check-abi-libdfp
FAIL: test-left_justify
Summary of test results:
      2 FAIL
    163 PASS
make: *** [tests] Error 1
```

**Comment by pdrocaldeira** [comment]
I authorize @tuliom to use my signed-off as he already did on my previous patch.

File: tests/test-left_justify.c [hunk 7]
```diff
@@ -169,12 +178,19 @@ d128_type d128[] =
...
-  /* __get_digits_d128 returns +0000000000000000000000000000000001e-6176 for a NaN data. */
-  {__LINE__, 0.0000000000000001E-6140DL, "+0000000000000000000000000000000001e-6176", "%s"},
-  {__LINE__, 0.0000000000000100E-6143DL, "+0000000000000000000000000000000001e-6176", "%s"},
-  {__LINE__, 0.0000000000001000E-6143DL, "+0000000000000000000000000000000001e-6176", "%s"},
-  {__LINE__, 0.0000000000000001E-6143DL, "+0000000000000000000000000000000001e-6176", "%s"},
+
+  {__LINE__, 0.0000000000000001E-6140DL, "+0000000000000100000000000000000000e-6176", "%s"},
+  {__LINE__, 0.0000000000000100E-6143DL, "+0000000000000010000000000000000000e-6176", "%s"},
+  {__LINE__, 0.0000000000001000E-6143DL, "+0000000000000100000000000000000000e-6176", "%s"},
+  {__LINE__, 0.0000000000000001E-6143DL, "+0000000000000000100000000000000000e-6176", "%s"},
```
Comment (tuliom) [inline tests/test-left_justify.c]: These changes are breaking x86_64.

**Comment by tuliom** [comment]
I've updated this PR with a version of the patch that solves most of the issues I pointed out.
However, this is still failing in 2 tests on x64_64:
test-left_justify
test-nextdown-d128

Reproduced with GCC 9.3.

**Comment by tuliom** [comment]
I've just fixed the issue on x86_64. There was a mistake on how `left_justifyd128` was computing the exponent of `y`. This error was already documented in the source code, but there was a minor mistake hiding it.

File: tests/test-left_justify.c (hunk 7 above)
Comment (pmur) [reply tests/test-left_justify.c:181]: And this seem fixed in the latest patchset.
